sdk/eventhub/azure-eventhub/azure/eventhub/_common.py

```diff
@@ -154,22 +171,52 @@ def __str__(self):
         return event_str
 
     @classmethod
-    def _from_message(cls, message):
-        # type: (Message) -> EventData
+    def _from_message(cls, message, raw_amqp_message=None):
+        # type: (Message, Optional[AmqpAnnotatedMessage]) -> EventData
         """Internal use only.
 
-        Creates an EventData object from a raw uamqp message.
+        Creates an EventData object from a raw uamqp message and, if provided, AmqpAnnotatedMessage.
 
         :param ~uamqp.Message message: A received uamqp message.
+        :param ~azure.eventhub.amqp.AmqpAnnotatedMessage message: An amqp annotated message.
         :rtype: ~azure.eventhub.EventData
         """
         event_data = cls(body="")
-        event_data.message = message
+        if raw_amqp_message and message:
+            event_data._raw_amqp_message = raw_amqp_message
+            event_data.message = message
+        elif message:
+            event_data._raw_amqp_message = AmqpAnnotatedMessage(message=message)
+            event_data.message = message
         return event_data
 
     def _encode_message(self):
         # type: () -> bytes
-        return self.message.encode_message()
+        # pylint: disable=protected-access
+        return self._raw_amqp_message._message.encode_message()
+
+    def _decode_non_data_body_as_str(self, encoding="UTF-8"):
+        # type: (str) -> str
+        # pylint: disable=protected-access
+        self.raw_amqp_message._message._body._encoding = encoding
```

Comment:
I don't think we should mutate the _encoding attribute of the _body, it will affect the following up decode/encode operations on the message.

but I should ask first, why you want to update the `_encoding` attribute to the _body?

Reply:
good point! I was thinking about whether to store the original encoding and reset it. The reason I update the encoding is because when we call `str(body)` for sequence body, it uses `body._encoding` to decode. Do you think this is needed @yunhaoling?

Reply:
updated to no longer mutate _encoding